PWGHF/HFL/Tasks/taskSingleMuonReaderAssoc.cxx

```diff
@@ -0,0 +1,193 @@
+    }
+    for (const auto& assoc : assocs) {
+      auto muon = assoc.template reducedmuon_as<TMuons>();
+      if (muon.trackType() != trkType) {
+        continue;
+      }
+
+      const auto eta(muon.eta()), pDca(muon.pDca());
+      const auto rAbs(muon.rAtAbsorberEnd());
+      const auto dcaXY(RecoDecay::sqrtSumOfSquares(muon.fwdDcaX(), muon.fwdDcaY()));
+      const auto pt(muon.pt());
+      const auto charge(muon.sign());
+      const auto chi2(muon.chi2MatchMCHMFT());
+
+      if ((eta >= etaMax) || (eta < etaMin)) {
+        continue;
+      }
+
+      if ((rAbs >= rAbsMax) || (rAbs < rAbsMin)) {
+        continue;
+      }
+      if (pDca >= pDcaMax) {
+        continue;
+      }
+
+      // histograms after acceptance cuts
+      if (muon.has_matchMCHTrack()) {
+        auto muonType3 = muon.template matchMCHTrack_as<TMuons>();
+        auto Dpt = muonType3.pt() - pt;
+
+        singleMuon(pt, dcaXY, Dpt, chi2);
+        registry.fill(HIST("hMuAfterCuts"), pt, eta, dcaXY, rAbs, charge, chi2, Dpt);
+      }
+    }
+  }
+  template <typename TCollisions, typename TMuons>
+  void runMuonSelMcAssoc(ReducedMuonsAssoc const& assocs, TCollisions const& collisions, TMuons const&)
+  {
+    for (const auto& collision : collisions) {
+      registry.fill(HIST("hVtxZ"), collision.posZ());
+    }
+    for (const auto& assoc : assocs) {
+      auto muon = assoc.template reducedmuon_as<TMuons>();
+      if (muon.trackType() != trkType) {
+        continue;
+      }
+
+      const auto eta(muon.eta()), pDca(muon.pDca());
+      const auto rAbs(muon.rAtAbsorberEnd());
+      const auto dcaXY(RecoDecay::sqrtSumOfSquares(muon.fwdDcaX(), muon.fwdDcaY()));
+      const auto pt(muon.pt());
+      const auto charge(muon.sign());
+      const auto chi2(muon.chi2MatchMCHMFT());
+
+      if ((eta >= etaMax) || (eta < etaMin)) {
+        continue;
+      }
+
+      if ((rAbs >= rAbsMax) || (rAbs < rAbsMin)) {
+        continue;
+      }
+      if (pDca >= pDcaMax) {
+        continue;
+      }
+
+      // histograms after acceptance cuts
+      if (muon.has_matchMCHTrack()) {
+        auto muonType3 = muon.template matchMCHTrack_as<TMuons>();
+        auto Dpt = muonType3.pt() - pt;
+
+        singleMuon(pt, dcaXY, Dpt, chi2);
+        registry.fill(HIST("hMuAfterCuts"), pt, eta, dcaXY, rAbs, charge, chi2, Dpt);
+        if (muon.mcMask() == 0) {
+          registry.fill(HIST("hMuAfterCutsTrue"), pt, eta, dcaXY, rAbs, charge, chi2, Dpt);
+        }
+        if (muon.mcMask() == 128) {
+          registry.fill(HIST("hMuAfterCutsFake"), pt, eta, dcaXY, rAbs, charge, chi2, Dpt);
+        }
+      }
+    }
+  }
+  void processMuon(ReducedMuonsAssoc const& assocs, MyCollisions const& collisions, MyMuons const& muons)
+  {
+    runMuonSelAssoc(assocs, collisions, muons);
+  }
+  void processMuonMc(ReducedMuonsAssoc const& assocs, MyCollisions const& collisions, MyMcMuons const& muons)
+  {
+    runMuonSelMcAssoc(assocs, collisions, muons);
+  }
```

Comment:
Why do you define the `runMuonSel...` methods instead of putting the code directly in the process function?

Reply:
Just in case we need different templates in the future